### ISSUE TITLE
Fix FP S4325 (`no-unnecessary-type-assertion`): `Object.values()` result

### ISSUE THREAD

**Issue by vilchik-elena** [SonarSource/SonarJS#3235]
```ts
export function isPositive(x: number) {
  return x > 0;
}

export type PositiveNumberRecord<T extends string | number | symbol> =
  Record<T, number> & { __type: 'positiveNumberRecord' };

export const isPositiveNumberRecord = <T extends string | number | symbol>
  (value: Record<T, number> | null | undefined): value is PositiveNumberRecord<T> => {

    if (!value) {
      return false;
    }

    const values = Object.values(value) as number[];  // FP is here

    return values.every(isPositive);
};
```

Reported in https://community.sonarsource.com/t/sonarlint-says-cast-is-unnecessary/64340

**Comment by victor-diez-sonarsource** [comment]
this comes from https://github.com/typescript-eslint/typescript-eslint/blob/dc58ff5da99989510fdbbe5575a31acd320b1808/packages/eslint-plugin/src/rules/no-unnecessary-type-assertion.ts#L257

seems the typechecker reports type Object for both

```
Object.values(value)
```
and 
```
Object.values(value) as number[];
```

**Comment by ilia-kebets-sonarsource** [comment]
The rule is complicated to implement since `TypeChecker. getTypeAtLocation(node)` returns `Object` for variables of type `Array`. There should be another method to find the exact type since VS code gives it when hovering.
The rule has many open issues with PRs on the repo, I will take a look into this: https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+is%3Aopen+no-unnecessary-type-assertion

**Comment by ilia-kebets-sonarsource** [comment]
After trying out multiple methods from the TypeChecker, we didn't manage to differentiate the nodes:
- TSasExpression
- TSasExpression.expression
- TSasExpression.typeAnnotation

We have tried the following methods (tc is a getTypeChecker()`):
- `tc.isTypeAssignableTo()`
- `tc.getTypeFromTypeNode()`
- `tc.getWidenedType()`
- `tc.getApparentType()`

Investigating [open issues for the rule on the typescript-eslint repo](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+is%3Aopen+no-unnecessary-type-assertion) didn't allow to find a similar issue.

**Comment by ilia-kebets-sonarsource** [comment]
I have opened a ticket on https://github.com/typescript-eslint/typescript-eslint/issues/5808

**Comment by ilia-kebets-sonarsource** [comment]
As mentioned [here](https://github.com/typescript-eslint/typescript-eslint/issues/5808#issuecomment-1274890635):

```
The workaround is to use the generic directly instead (Object.values<number>()), or use a disable comment.
```
